htmresearch/algorithms/column_pooler.py

```diff
@@ -174,16 +192,127 @@ def compute(self, feedforwardInput=(), lateralInputs=(),
 
     @param  learn (bool)
             If True, we are learning a new object
-    """
 
+    @param predictedInput (sequence)
+           Sorted indices of active feedforward input bits
+    """
     if feedforwardGrowthCandidates is None:
       feedforwardGrowthCandidates = feedforwardInput
 
-    if learn:
+    if not learn:
+      self._computeInferenceMode(feedforwardInput, lateralInputs)
+
+    elif not self.onlineLearning:
       self._computeLearningMode(feedforwardInput, lateralInputs,
                                 feedforwardGrowthCandidates)
     else:
-      self._computeInferenceMode(feedforwardInput, lateralInputs)
+      if (predictedInput is not None and
+          len(predictedInput) > self.predictedInhibitionThreshold):
```

Comment:
I'm surprised that this code isn't saying "Alright, there was a prediction, so I'm going to require all active cells in the pooling layer to have feedforward support if they want to stay active." Instead, if the input doesn't match the prediction, it's as if there's no feedforward input. So the inertia will keep many previously active cells active. Then we activate a few other random cells, then learn on that SDR?

For example, if the sequence ABCDE is known, and then you're seeing APQRS for the first time, upon "P" I'd expect this code to deactivate the "ABCDE" sequence SDR, then activate a random "APQRS" SDR. Instead, depending on the inertia and learningTolerance parameters, it will create a "APQRS" SDR that is similar to the "ABCDE" SDR?

(Aside: I don't know how a random "APQRS" SDR would ever learn "A". For object recognition, that's fine, because you can always revisit locations, but for sequence learning I don't know how you would ever prepend elements to a new sequence, other than by using a new form of learning that goes multiple timesteps backward.)